### PR TITLE
[ISSUE #2435] Solve the problem that DefaultMQProducer#request() sends messages and waits for timeout synchronously

### PR DESCRIPTION
**Pull request by li-xiao-shuang**
fix : #2435 

1.  Add the default namesrvAddr to the RequestProducer example
2.  Solve the synchronization wait timeout problem because the semaphore does not decrease after DefaultMQProducer#request() message is sent successfully
3. Solve the problem that the callback method is not called after the DefaultMQProducer#request() message is sent successfully

### CODE REVIEW

File: client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java [hunk 1]
```diff
@@ -1416,11 +1417,13 @@ public void request(Message msg, final RequestCallback requestCallback, long tim
...
             public void onSuccess(SendResult sendResult) {
                 requestResponseFuture.setSendRequestOk(true);
+                requestResponseFuture.executeRequestCallback();
             }
 
             @Override
             public void onException(Throwable e) {
                 requestResponseFuture.setCause(e);
+                requestResponseFuture.executeRequestCallback();
                 requestFail(correlationId);
```
Comment (hzh0425) [inline client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java:1427]: ' requestResponseFuture.executeRequestCallback' has been called in  'requestFail(correlationId)'
you don't need to call it again.

**Review by hzh0425 — COMMENTED** [review]
According to the above logic, should 'putResponseMessage' also be called in line 1443? and the following request has the same problem?
image

File: client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java (hunk 1 above)
Comment (li-xiao-shuang) [reply client/src/main/java/org/apache/rocketmq/client/impl/producer/DefaultMQProducerImpl.java]: done

**Comment by li-xiao-shuang** [comment]
> According to the above logic, should 'putResponseMessage' also be called in line 1443? and the following request has the same problem? image

Yes, I will deal with this separately. The other interfaces are special. I am still familiar with MQ

**Comment by hzh0425** [comment]
> > According to the above logic, should 'putResponseMessage' also be called in line 1443? and the following request has the same problem? image
> 
> Yes, I will deal with this separately. The other interfaces are special. I am still familiar with MQ

Okay, nice try

**Review by duhenglucky — APPROVED** [review]

**Comment by martincao12** [comment]
这么修改相当于在执行sendCallback的onSuccess的时候就立即执行了requestResponseFuture的putResponseMessage或者executeRequestCallback，这俩不是应该在收到replyMsg之后再执行吗，即ClientRemotingProcessor里processReplyMessage做的事情？看Issue #2435里的报错，是因为broker reply msg的时候，request请求的producer已经unregister了。

**Pull request merged**